src/nxt_http_return.c

```diff
@@ -38,30 +36,9 @@ nxt_http_return_init(nxt_mp_t *mp, nxt_http_action_t *action,
 
     conf->status = nxt_conf_get_number(acf->ret);
 
-    if (acf->location.length > 0) {
-        if (nxt_is_complex_uri_encoded(acf->location.start,
-                                       acf->location.length))
-        {
-            loc = nxt_str_dup(mp, &conf->location, &acf->location);
-            if (nxt_slow_path(loc == NULL)) {
-                return NXT_ERROR;
-            }
-
-        } else {
-            loc = &conf->location;
-
-            encode = nxt_encode_complex_uri(NULL, acf->location.start,
-                                            acf->location.length);
-            loc->length = acf->location.length + encode * 2;
-
-            loc->start = nxt_mp_nget(mp, loc->length);
-            if (nxt_slow_path(loc->start == NULL)) {
-                return NXT_ERROR;
-            }
-
-            nxt_encode_complex_uri(loc->start, acf->location.start,
-                                   acf->location.length);
-        }
```

Comment:
Previously this encoding was done once during reconfiguration phase. Now it happens on every request.

Common approach here is to move this code into a separate function and then do encoding (if needed) on the reconfiguration phase if option doesn't contain variables, or do it during in runtime after interpolation.

Reply:
Thanks.  I was concerned that it was very bad for performance the way I did it, but I wasn't sure how to do it better.  Which is the most similar code to look at?  share?  Or is there some simpler case?

Reply:
"chroot" option is a similar example.

Reply:
Thanks for the tips.  Now I truly understand how it works and am very happy with the code!  I used a mix of "pass", "share", and "chroot" tounderstand it :).

Reply:
Based on the [RFC 7231](https://datatracker.ietf.org/doc/html/rfc7231#section-6.4.2) the server SHOULD return a location header and the Client MAY use it to redirect the user. I follow your idea of NOT sending an empty HTTP Location header but do not send any.

Reply:
Thanks for finding the RFC.  So it allows us to not send an HTTP Location header.  That's good.

But I still wonder the following:

Is there any real-world scenario where a server might have an empty Location string?
We're optimizing for it, and even if it just adds a couple of operations in the commit case (to check if the string is empty a couple of times), the code would be cleaner without that optimization.

However, if you think that it might be useful to have an empty location string for some reason, yes, let's optimise.

Reply:
Location must be a URI and a URI may not be empty. In the case of a 201 response, an empty Location header has the same meaning as no Location header.

Reply:
> Location must be a URI and a URI may not be empty.

That makes a lot of sense.

Reply:
Should that path be `nxt_slow_path()`?

Reply:
Nope, just the same as before.

Reply:
The obsolete RFC 2616 defined the Location to be an absolute URI <https://www.rfc-editor.org/rfc/rfc2616#section-14.30>, which as you say, cannot be an empty string.

However, the current RFC allows the Location to be a relative URI <https://www.rfc-editor.org/rfc/rfc7231#section-7.1.2>, and a relative URI might be the empty string.

Reply:
I'll mark this as resolved, since it won't be changed in this patch set.  I'll start a separate discussion for it.

Reply:
https://github.com/nginx/unit/issues/674